Fix and tests added DAFFODIL-2873 zoned ebcdic bug

mbeckerle opened the pull request:
DAFFODIL-2873

mbeckerle commented on daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml:
```diff
@@ -0,0 +1,96 @@
+
+  </tdml:defineSchema>
+
+  <tdml:parserTestCase name="ZonedEBCDICLeadingSign" root="money7" model="ebcdic1"
```
tdml: prefixes can all be removed.

stevedlawrence commented on daffodil-test/src/test/resources/org/apache/daffodil/section13/zoned/zoned2.tdml:
```diff
@@ -0,0 +1,96 @@
+  <parserTestCase name="ZonedEBCDICLeadingSign" root="money7" model="ebcdic1"
+    description="Test of a negative number with ebcdic charset, leading overpunched sign of R should be -9">
+    <document>
+      <documentPart type="text" encoding="ebcdic-cp-us">R99999999</documentPart>
```
It's there any value in defining the document part in bytes instead of text with ebcdic-cp-us encoding? That would show that the issue doesn't have anything to do with ebcdic encoding in the documentPartt? I feel like that isn't used very often so there's a chance the bug is in that?

stevedlawrence reviewed (approved):
+1 needs  `sbt scalafmtAll` to quiet the linter.

mbeckerle commented on daffodil-core/src/main/scala/org/apache/daffodil/core/dsom/RuntimePropertyMixins.scala:
```diff
@@ -123,7 +123,7 @@ trait TermRuntimeValuedPropertiesMixin
   }
 
-  final lazy val charsetEv = {
+  final lazy val charsetEv: CharsetEv = {
```
Revert

mbeckerle commented on daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:
```diff
@@ -700,6 +701,30 @@ trait ElementBaseGrammarMixin
+      .map { bcs: BitsCharset =>
+        val nom = bcs.name.toUpperCase()
+        val res = (nom == "IBM037") ||
+          nom.startsWith("EBCDIC")
```
This may not be sufficient to recognize EBCDIC. There are numrous IBM code pages that are EBCDIC besides IBM037. There are ones for the international variants of EBCDIC. All of them do the exact same thing for zoned numbers. 

We need something that will work regardless of the aliases vs. the "official" name. If the name, or any alias contains "ebcdic" (case insensitive) then it is an ebcdic variant.

stevedlawrence replied:
Should we add an `isEBCDIC` member to `BitsCharset` that defaults to `false`, and any EBCDIC encodings we implement would override that to `true`? Then this just becomes

```scala
charsetEv.optConstant.map { _.isEBCDIC }.getOrElse(false)
```

Also, doing some grepping for EBCDIC I see we also have a `Misc.isAsciiBased` function. That's only used for dumping debug data and works on Java charsets instead of our BitsCharset, but maybe it provides extra motivation that we might eventually want to have other `is*` members so we can ask questions about our supported charsets without having to make guesses based on the name.

stevedlawrence commented on daffodil-core/src/main/scala/org/apache/daffodil/core/grammar/ElementBaseGrammarMixin.scala:
```diff
@@ -700,6 +701,30 @@ trait ElementBaseGrammarMixin
+  lazy val optTextZonedSignStyle = {
+    if (isKnownEBCDICEncoding) None
+    else Some(textZonedSignStyle)
+  }
```
This doesn't do the right thing if the charset is non-constant be evaluates to EBCIDC at runtime. Is that something we care about? Does it make sense to convert this to a evaluatable?

stevedlawrence replied:
But if the charset evaluates to EBCDIC at runtime we'll still use textZonedSignStyle. So in that case we still need the property at compile time (like you have) but we need to ignore it at runtime.

So maybe the parser/unparsers need to have the charsetEv passed in, and if at runtime if it's non-constant and evaluates to EBCDIC then it ignores `optTextZonedSignStyle`?

mbeckerle replied:
Correct. That is what happens. So this optional thing is just about not reqiring what is an ascii-oriented property at all in the case where it is known for certain to be EBCDIC.

stevedlawrence commented on daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala:
```diff
@@ -386,6 +386,25 @@ object DecimalUtils {
+    else if ((digit >= 'J') && (digit <= 'R')) // negative 1-9
+      (digit - 'J' + 1, true)
+    else
+      throw new NumberFormatException("Invalid zoned digit: " + digit)
```
The DFDL spec says this about `textNumberRep`:
> In EBCDIC-based encodings, code points 0xC0 to 0xC9 or 0xF0 to 0xF9 represent a positive sign and digits 0 to 9 (these byte ranges correspond typically to characters '{ABCDEFGHI'* or '0123456789'), and code points 0xD0 to 0xD9 or 0xB0 to 0xB9 represent a negative sign and digits 0 to 9 (these byte ranges correspond typically to characters '}JKLMNOPQR' or  '^£¥·©§¶¼½¾ ' ). On parsing both ranges are accepted. On unparsing the range 0xC0 to 0xC9 are produced for positive signs and the range 0xD0 to 0xD9 are produced for negative signs.

Should this function also support `{ABCDEFGHI` for positive numbers and `^£¥·©§¶¼½¾ ` for negatives?

mbeckerle replied:
Correct. Fixed.

stevedlawrence commented on daffodil-lib/src/main/scala/org/apache/daffodil/lib/util/DecimalUtils.scala:
```diff
@@ -417,6 +436,25 @@ object DecimalUtils {
+   * are mapped to "}JKLMNOPQR".
+   * @param digit - the digit, as a char '0' to '9'
+   * @param positive - true if positive, false if negative
+   * @return the character needed to represent this character as an overpunched sign digit.
```
From the spec:

> In EBCDIC-based encodings, code points 0xC0 to 0xC9 or 0xF0 to 0xF9 represent a positive sign and digits 0 to 9 (these byte ranges correspond typically to characters '{ABCDEFGHI' or '0123456789'), and code points 0xD0 to 0xD9 or 0xB0 to 0xB9 represent a negative sign and digits 0 to 9 (these byte ranges correspond typically to characters '}JKLMNOPQR' or  '^£¥·©§¶¼½¾ ' ). On parsing both ranges are accepted. On unparsing the range 0xC0 to 0xC9 are produced for positive signs and the range 0xD0 to 0xD9 are produced for negative signs.

This makes it sound like on unparsing the positive 0-9 digits should be unparsed as `{ABCDFGHI` instead of 0-9?

mbeckerle replied:
You are correct. Fixed.

stevedlawrence commented on daffodil-lib/src/test/scala/org/apache/daffodil/lib/util/TestDecimalUtils.scala:
```diff
@@ -1846,733 +1989,1620 @@ class TestDecimalUtils {
       "",
     )
   }
+
```
Same question as above, should the positives be `{` and `A-I` instead of `0-9`? And do we need tests for the other range of positive/negative values?

mbeckerle replied:
You are correct. Fixed and tests added.

stevedlawrence reviewed (approved):
+1

mbeckerle commented:
I believe this now addresses all the comments by @stevedlawrence.

Needs a second reviewer. @jadams-tresys can you review this? I believe the ascii zoned stuff was first done by you.

jadams-tresys commented:
+1

Looks good to me.  Reading the Wikipedia page I'm not surprised I didn't get this right on the first try.  This section was a fun read though: https://en.wikipedia.org/wiki/EBCDIC#Criticism_and_humor
